### PR TITLE
refactor: refactor to promises

### LINKED ISSUE #3: Implement promises instead of callbacks

### PR DESCRIPTION
**Pull request by Lehren**
Refactor tests and implementation to promises.

Resolve #3

### CODE REVIEW

**Comment by kvdb** [comment]
Also, please use https://help.github.com/articles/closing-issues-via-commit-messages/

File: administrative-sdk/basicAuth.js [hunk 1]
```diff
@@ -38,55 +38,32 @@ class BasicAuth {
...
+          // Credentials are only supplied when generated by the backend.
+          if (data.credentials) {
+            self.credentials = data.credentials;
+          }
+          resolve(self);
+        })
+        .catch(function(error) {
+          reject(error);
+        });
+    });
```
Comment (remcohaszing) [inline administrative-sdk/basicAuth.js]: One of the great advantages is to prevent callback hell. This can be done by chaining them. This piece of code could be greatly simplified

```js
return connection._secureAjaxPost(url, JSON.stringify(this))
  .then(data => {
    this.principal = data.principal;
    this.created = new Date(data.created);
    this.updated = new Date(data.updated);
    // Credentials are only supplied when generated by the backend.
    if (data.credentials) {
      this.credentials = data.credentials;
    }
    return this;
  });
```

Also note that `this` refers to the `basicAuth` instance, because an arrow function is used. There's barely any need to ever reassign `this` to `self` anymore.

This comment applies to many of the same types of refactor I've seen so far.

Comment (Lehren) [inline administrative-sdk/basicAuth.js]: Applied changes in latest commit

File: administrative-sdk/choiceRecognition.js [hunk 2]
```diff
@@ -110,160 +94,148 @@ class ChoiceRecognition {
...
-        var analysis = new PronunciationAnalysis(
-          challenge.id, data.studentId, data.id,
-          new Date(data.created), new Date(data.updated),
-          connection.addAccessToken(data.audioUrl));
-        ecb(analysis, data.message);
-      }
-    };
-
-    // Validate required domain model.
+  startStreamingChoiceRecognition(connection, challenge, recorder, trim) {
```
Comment (remcohaszing) [inline administrative-sdk/choiceRecognition.js]: This method is pretty hard to follow... but I don't think it's this change is to blame.

File: administrative-sdk/choiceRecognition.js [hunk 3]
```diff
@@ -273,91 +245,68 @@ class ChoiceRecognition {
    * @param {Sdk~getChoiceRecognitionCallback} [cb] The callback that handles the response.
    * @param {Sdk~getChoiceRecognitionErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/choiceRecognition.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/choiceRecognition.js [hunk 4]
```diff
@@ -273,91 +245,68 @@ class ChoiceRecognition {
...
-      }
-    };
+  }
 
+  /**
+   * List all choice recognitions in a specific choice challenge.
+   *
+   * @param {ChoiceChallenge} challenge Specify a choice challenge to list speech recognitions for.
+   * @param {Sdk~listChoiceRecognitionsCallback} cb The callback that handles the response.
+   * @param {Sdk~listChoiceRecognitionsErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/choiceRecognition.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 5]
```diff
@@ -142,8 +142,8 @@ class Connection {
    * @param {callback} [cb] The callback that handles the response.
    * @param {callback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 6]
```diff
@@ -154,8 +154,8 @@ class Connection {
...
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 7]
```diff
@@ -165,8 +165,8 @@ class Connection {
...
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 8]
```diff
@@ -192,32 +193,34 @@ class Connection {
...
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 9]
```diff
@@ -192,32 +193,34 @@ class Connection {
...
    * @param {string} [auth] The authorization header value to pass along with the request.
    */
-  _ajaxGet(url, cb, ecb, auth) {
-    var self = this;
+  _ajaxGet(url, auth) {
     var request = new XMLHttpRequest();
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Part of converting to promises is to rewrite XMLHttpRequest to the new [fetch](https://github.com/github/fetch) API.

File: administrative-sdk/connection.js [hunk 10]
```diff
@@ -229,46 +232,44 @@ class Connection {
    * @param {callback} [ecb] The callback that handles the error response.
    * @param {string} [auth] The authorization header value to pass along with the request.
    */
-  _ajaxPost(url, formdata, cb, ecb, auth) {
-    var self = this;
-    var request = new XMLHttpRequest();
+  _ajaxPost(url, formdata, auth) {
+    var request = new XMLHttpRequest({mozSystem: true});
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Part of converting to promises is to rewrite XMLHttpRequest to the new fetch API.

File: administrative-sdk/connection.js [hunk 11]
```diff
@@ -229,46 +232,44 @@ class Connection {
    * @param {callback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 12]
```diff
@@ -279,36 +280,36 @@ class Connection {
...
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/connection.js [hunk 13]
```diff
@@ -279,36 +280,36 @@ class Connection {
...
    * @param {string} [auth] The authorization header value to pass along with the request.
    */
-  _ajaxDelete(url, cb, ecb, auth) {
-    var self = this;
+  _ajaxDelete(url, auth) {
     var request = new XMLHttpRequest();
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Part of converting to promises is to rewrite XMLHttpRequest to the new fetch API.

File: administrative-sdk/organisation.js [hunk 14]
```diff
@@ -20,67 +20,24 @@ class Organisation {
...
-   * @param {its.Organisation} organisation Organisation domain model instance with unapplied changes.
-   */
-  organisationCreatedErrorCallback(errors, organisation) {}
-
-  /**
    * Create an organisation.
    *
    * @param {its.Organisation} organisation An organisation domain model instance.
    * @param {Sdk~organisationCreatedCallback} [cb] The callback that handles the response.
    * @param {Sdk~organisationCreatedErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/organisation.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/organisation.js [hunk 15]
```diff
@@ -89,56 +46,38 @@ class Organisation {
    * @param {Sdk~getCallback} [cb] The callback that handles the response.
    * @param {Sdk~getErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/organisation.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/organisation.js [hunk 16]
```diff
@@ -89,56 +46,38 @@ class Organisation {
...
+        organisation.updated = new Date(data.updated);
+        return (organisation);
+      });
+  }
 
   /**
    * List all organisations in the organisation.
    *
    * @param {Sdk~listCallback} cb The callback that handles the response.
    * @param {Sdk~listErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/organisation.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/pronunciationAnalysis.js [hunk 17]
```diff
@@ -9,13 +9,16 @@
...
 /**
  * @class WordChunk
  *
  * @member {string} graphemes The graphemes this chunk consists of.
- * @member {float} score The audio is scored per grapheme and consists of several measurements. 0 would be bad, 1 the perfect score.
- * @member {string} verdict `bad` when the score is below 0.4, `moderate` when equal to 0.4 or between 0.4 and 0.6. `good` when the score is 0.6 or above.
+ * @member {float} score The audio is scored per grapheme and consists of several measurements. 0 would be bad, 1 the
+ *   perfect score.
+ * @member {string} verdict `bad` when the score is below 0.4, `moderate` when equal to 0.4 or between 0.4 and 0.6.
+ *   `good` when the score is 0.6 or above.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Why have these lines changed? Please avoid making unrelated changes.

File: administrative-sdk/pronunciationAnalysis.js [hunk 18]
```diff
@@ -24,8 +27,10 @@ class WordChunk {
    *
    * @constructor
    * @param {string} graphemes The graphemes this chunk consists of.
-   * @param {float} score The audio is scored per grapheme and consists of several measurements. 0 would be bad, 1 the perfect score.
-   * @param {string} verdict `bad` when the score is below 0.4, `moderate` when equal to 0.4 or between 0.4 and 0.6. `good` when the score is 0.6 or above.
+   * @param {float} score The audio is scored per grapheme and consists of several measurements. 0 would be bad, 1 the
+   *   perfect score.
+   * @param {string} verdict `bad` when the score is below 0.4, `moderate` when equal to 0.4 or between 0.4 and 0.6.
+   *   `good` when the score is 0.6 or above.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Why have these lines changed? Please avoid making unrelated changes.

File: administrative-sdk/pronunciationAnalysis.js [hunk 19]
```diff
@@ -59,18 +64,23 @@ class Word {
  * @class Phoneme
  *
  * @member {string} ipa The pronunciation of the grapheme(s) indicated as International Phonetic Alphabet (IPA).
- * @member {float} score The audio is scored per phoneme and consists of several measurements. 0 would be bad, 1 the perfect score.
- * @member {string} bad when the score is below 0.4, moderate when equal to 0.4 or between 0.4 and 0.6. good when the score is 0.6 or above.
+ * @member {float} score The audio is scored per phoneme and consists of several measurements. 0 would be bad, 1 the
+ *   perfect score.
+ * @member {string} bad when the score is below 0.4, moderate when equal to 0.4 or between 0.4 and 0.6. good when the
+ *   score is 0.6 or above.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Why have these lines changed? Please avoid making unrelated changes.

File: administrative-sdk/pronunciationAnalysis.js [hunk 20]
```diff
@@ -59,18 +64,23 @@ class Word {
...
    * @param {string} ipa The pronunciation of the grapheme(s) indicated as International Phonetic Alphabet (IPA).
-   * @param {float} score The audio is scored per phoneme and consists of several measurements. 0 would be bad, 1 the perfect score.
-   * @param {float} confidenceScore This value provides a reliable prediction that the pronounced phoneme is actually the phoneme that is supposed to be pronounced. There is no absolute scale defined yet.
-   * @param {string} verdict bad when the score is below 0.4, moderate when equal to 0.4 or between 0.4 and 0.6. good when the score is 0.6 or above.
+   * @param {float} score The audio is scored per phoneme and consists of several measurements. 0 would be bad, 1 the
+   *   perfect score.
+   * @param {float} confidenceScore This value provides a reliable prediction that the pronounced phoneme is actually
+   *   the phoneme that is supposed to be pronounced. There is no absolute scale defined yet.
+   * @param {string} verdict bad when the score is below 0.4, moderate when equal to 0.4 or between 0.4 and 0.6. good
+   *   when the score is 0.6 or above.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Why have these lines changed? Please avoid making unrelated changes.

File: administrative-sdk/pronunciationAnalysis.js [hunk 21]
```diff
@@ -92,7 +102,8 @@ class Phoneme {
  * @member {blob} audio The recorded audio fragment.
  * @member {string} audioUrl The audio fragment as streaming audio link.
  * @member {number} score The average score of all phonemes grading the entire attempt.
- * @member {float} confidenceScore This value provides a reliable prediction that the pronounced phonemes are actually the phonemes that are supposed to be pronounced. There is no absolute scale defined yet.
+ * @member {float} confidenceScore This value provides a reliable prediction that the pronounced phonemes are actually
+ *   the phonemes that are supposed to be pronounced. There is no absolute scale defined yet.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Why have these lines changed? Please avoid making unrelated changes.

File: administrative-sdk/pronunciationAnalysis.js [hunk 22]
```diff
@@ -230,246 +215,169 @@ class PronunciationAnalysis {
...
    * @param {its.PronunciationChallenge} challenge The pronunciation challenge to perform.
    * @param {its.AudioRecorder} recorder The audio recorder to extract audio from.
-   * @param {Sdk~pronunciationAnalysisPreparedCallback} [preparedCb] The callback that signals server is prepared for receiving data.
+   * @param {Sdk~pronunciationAnalysisPreparedCallback} [preparedCb] The callback that signals server is prepared for
+   *   receiving data.
    * @param {Sdk~pronunciationAnalysisCreatedCallback} [cb] The callback that handles the response.
    * @param {Sdk~pronunciationAnalysisCreatedErrorCallback} [ecb] The callback that handles the error response.
-   * @param {Sdk~pronunciationAnalysisProgressCallback} [progressCb] The callback that handles the intermediate results.
+   * @param {Sdk~pronunciationAnalysisProgressCallback} [progressCb] The callback that handles the intermediate
+   *   results.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/pronunciationAnalysis.js [hunk 23]
```diff
@@ -9,13 +9,16 @@
 const Student = require('../administrative-sdk/student').Student;
 const Base64Utils = require('./base64Utils').Base64Utils;
 const Connection = require('../administrative-sdk/connection').Connection;
+var when = require('autobahn').when;
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Why not just use `require('when');`?

Also, use a const.

File: administrative-sdk/pronunciationAnalysis.js [hunk 24]
```diff
@@ -230,246 +215,169 @@ class PronunciationAnalysis {
...
-   */
-  getPronunciationAnalysisErrorCallback(errors) {}
-
-  /**
    * Get a pronunciation analysis in a pronunciation challenge.
    *
    * @param {PronunciationChallenge} challenge Specify a pronunciation challenge.
    * @param {string} analysisId Specify a pronunciation analysis identifier.
    * @param {Sdk~getPronunciationAnalysisCallback} [cb] The callback that handles the response.
    * @param {Sdk~getPronunciationAnalysisErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/pronunciationAnalysis.js [hunk 25]
```diff
@@ -481,27 +389,50 @@ class PronunciationAnalysis {
...
-    };
+  }
 
+  /**
+   * List all pronunciation analyses in a specific pronunciation challenge.
+   *
+   * @param {PronunciationChallenge} challenge Specify a pronunciation challenge to list speech recordings for.
+   * @param {Boolean} detailed Returns extra analysis metadata when true. false by default.
+   * @param {Sdk~listPronunciationAnalysesCallback} cb The callback that handles the response.
+   * @param {Sdk~listPronunciationAnalysesErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationAnalysis.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/pronunciationChallenge.js [hunk 26]
```diff
@@ -53,60 +53,22 @@ class PronunciationChallenge {
...
-   * @param {its.PronunciationChallenge} challenge Pronunciation challenge domain model instance with unapplied changes.
-   */
-  pronunciationChallengeCreatedErrorCallback(errors, challenge) {}
-
-  /**
    * Create a pronunciation challenge.
    *
    * @param {its.PronunciationChallenge} challenge A pronunciation challenge object.
    * @param {Sdk~pronunciationChallengeCreatedCallback} [cb] The callback that handles the response.
    * @param {Sdk~pronunciationChallengeCreatedErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationChallenge.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/pronunciationChallenge.js [hunk 27]
```diff
@@ -115,150 +77,86 @@ class PronunciationChallenge {
...
-   * @param {object[]} errors Array of errors.
-   */
-  listPronunciationChallengesErrorCallback(errors) {}
-
-  /**
    * List all pronunciation challenges in the organisation.
    *
    * @param {string} organisationId Specify an organisation identifier.
    * @param {Sdk~listPronunciationChallengesCallback} cb The callback that handles the response.
    * @param {Sdk~listPronunciationChallengesErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationChallenge.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/pronunciationChallenge.js [hunk 28]
```diff
@@ -115,150 +77,86 @@ class PronunciationChallenge {
...
-   * @param {its.PronunciationChallenge} challenge Pronunciation challenge domain model instance with unapplied changes.
-   */
-  pronunciationChallengeDeletedErrorCallback(errors, challenge) {}
-
-  /**
    * Delete a pronunciation challenge.
    *
    * @param {its.PronunciationChallenge} challenge A pronunciation challenge object.
    * @param {Sdk~pronunciationChallengeDeletedCallback} [cb] The callback that handles the response.
    * @param {Sdk~pronunciationChallengeDeletedErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/pronunciationChallenge.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/speechChallenge.js [hunk 29]
```diff
@@ -54,52 +54,16 @@ class SpeechChallenge {
...
-   * @param {its.SpeechChallenge} challenge Speech challenge domain model instance with unapplied changes.
-   */
-  speechChallengeCreatedErrorCallback(errors, challenge) {}
-
-  /**
    * Create a speech challenge.
    *
    * @param {its.SpeechChallenge} challenge A speech challenge object.
    * @param {Sdk~speechChallengeCreatedCallback} [cb] The callback that handles the response.
    * @param {Sdk~speechChallengeCreatedErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/speechChallenge.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/speechChallenge.js [hunk 30]
```diff
@@ -177,24 +115,22 @@ class SpeechChallenge {
    * @param {Sdk~listSpeechChallengesCallback} cb The callback that handles the response.
    * @param {Sdk~listSpeechChallengesErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/speechChallenge.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/speechRecording.js [hunk 31]
```diff
@@ -123,197 +102,155 @@ class SpeechRecording {
...
    * @param {its.SpeechChallenge} challenge The pronunciation challenge to perform.
    * @param {its.AudioRecorder} recorder The audio recorder to extract audio from.
-   * @param {Sdk~speechRecordingPreparedCallback} [preparedCb] The callback that signals server is prepared for receiving data.
-   * @param {Sdk~speechRecordingCreatedCallback} [cb] The callback that handles the response. The success outcome is returned as first parameter, whether the recording was forcedStopped due to timer timeout is returned as second parameter.
+   * @param {Sdk~speechRecordingPreparedCallback} [preparedCb] The callback that signals server is prepared for
+   *   receiving data.
+   * @param {Sdk~speechRecordingCreatedCallback} [cb] The callback that handles the response. The success outcome is
+   *   returned as first parameter, whether the recording was forcedStopped due to timer timeout is returned as second
+   *   parameter.
    * @param {Sdk~speechRecordingCreatedErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/speechRecording.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/speechRecording.js [hunk 32]
```diff
@@ -123,197 +102,155 @@ class SpeechRecording {
...
-   */
-  getSpeechRecordingErrorCallback(errors) {}
-
-  /**
    * Get a speech recording in a speech challenge.
    *
    * @param {SpeechChallenge} challenge Specify a speech challenge.
    * @param {string} recordingId Specify a speech recording identifier.
    * @param {Sdk~getSpeechRecordingCallback} [cb] The callback that handles the response.
    * @param {Sdk~getSpeechRecordingErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/speechRecording.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/speechRecording.js [hunk 33]
```diff
@@ -322,34 +259,31 @@ class SpeechRecording {
    * @param {Sdk~listSpeechRecordingsCallback} cb The callback that handles the response.
    * @param {Sdk~listSpeechRecordingsErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/speechRecording.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/student.js [hunk 34]
```diff
@@ -123,63 +80,41 @@ class Student {
    * @param {Sdk~getCallback} [cb] The callback that handles the response.
    * @param {Sdk~getErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/student.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

File: administrative-sdk/student.js [hunk 35]
```diff
@@ -123,63 +80,41 @@ class Student {
...
-   * @param {object[]} errors Array of errors.
-   */
-  studentListErrorCallback(errors) {}
-
-  /**
    * List all students in the organisation.
    *
    * @param {string} organisationId Specify an organisation identifier.
    * @param {Sdk~listCallback} cb The callback that handles the response.
    * @param {Sdk~listErrorCallback} [ecb] The callback that handles the error response.
```
Comment (remcohaszing) [inline administrative-sdk/student.js]: Missing documentation update. Callbacks are no longer taken and a promise is returned.

I think you're getting the idea now... `git grep` is your friend.

File: test/basicAuthSpec.js [hunk 36]
```diff
@@ -10,12 +10,12 @@
  expect,
  it,
  jasmine,
+ fail,
```
Comment (remcohaszing) [inline test/basicAuthSpec.js]: Is this really necessary? the `jasmine` env is set in `.eslintrc`.

File: administrative-sdk/connection.js [hunk 37]
```diff
@@ -194,33 +194,27 @@ class Connection {
...
+          .then(data => {
+            if (response.ok) {
+              return data;
+            }
+            throw data;
+          })
+      ))
+      .catch(error => {
+        throw error;
+      });
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: This can be simplified by chaining the promises, not nesting them. Also the catch seems to be unnecessary.

```js
return fetch(url, options)
  .then(response => response.json())
  .then(data => {
    if (response.ok) {
      return data;
    }
    throw data;
  });
```

File: administrative-sdk/connection.js [hunk 38]
```diff
@@ -233,43 +227,32 @@ class Connection {
...
-      };
-
-      request.send(formdata);
-    });
+            throw data;
+          })
+      ))
+      .catch(error => {
+        throw error;
+      });
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Same as in `_ajaxGet`. Although I do think some code could be shared between those two.

File: administrative-sdk/connection.js [hunk 39]
```diff
@@ -281,35 +264,27 @@ class Connection {
...
+          .then(data => {
+            if (response.ok) {
+              return data;
+            }
+            throw data;
+          })
+      ))
+      .catch(error => {
+        throw error;
+      });
```
Comment (remcohaszing) [inline administrative-sdk/connection.js]: Same as in `_ajaxGet`. Although I do think some code could be shared between those two.

File: administrative-sdk/basicAuth.js [hunk 40]
```diff
@@ -40,9 +40,7 @@ class BasicAuth {
   /**
    * Create a basic auth.
    *
-   * @param {its.BasicAuth} basicAuth A basic auth domain model instance.
-   * @param {Sdk~basicAuthCreatedCallback} [cb] The callback that handles the response.
-   * @param {Sdk~basicAuthCreatedErrorCallback} [ecb] The callback that handles the error response.
+   * @param {Connection} connection object to connect to.
```
Comment (remcohaszing) [inline administrative-sdk/basicAuth.js]: Capitalize the first word. `Object to connect to.`
